### ISSUE TITLE
Set human-readable names for the editor's nodes

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#1018]
**Describe the project you are working on:**

The Godot editor :slightly_smiling_face: 

**Describe the problem or limitation you are having in your project:**

When implementing new features in the editor, it's often useful to refer to the scene tree to see how things are organized. This can be done using `print_tree_pretty()` or with Zylann's excellent [EditorDebugger](https://github.com/Zylann/godot_editor_debugger_plugin) add-on.

Unfortunately, nodes don't have manually-assigned names, which makes this tree much less human-readable than it could be.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

Assigning names to nodes will make it easier for contributors to work on the editor. This will let them fix bugs and add new features more efficiently.

This can also be useful for add-on developers. That said, we don't make any guarantees as to node names not changing over time, so this should not be regarded as a reliable solution for `get_node()` in add-ons.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

This is already being done in some places, such as here:

https://github.com/godotengine/godot/blob/f54eda121d975127f68db624265271d39d8dabf3/editor/node_3d_editor_gizmos.cpp#L4257

This proposal is just a matter of doing the same for all nodes used in the editor. Names should use the PascalCase convention for consistency, unless the node name is being displayed in the UI somehow (e.g. for use in TabContainer).

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, as this is core editor functionality.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is core editor functionality.

**Comment by pycbouh** [comment]
Do you propose to name *every* node of the editor UI or are we going to determine some that are more valuable? I'm not sure that there is any practical benefit from the former.

**Comment by Calinou** [comment]
> Do you propose to name *every* node of the editor UI or are we going to determine some that are more valuable? I'm not sure that there is any practical benefit from the former.

I think the long term goal would be to name every node in the editor. Of course, we'll have to set priorities and start with top-level nodes first (docks, menus, …).

**Comment by Zylann** [comment]
There are many nodes that are only there for layout purposes (containers), naming them might be a useless hassle.

**Comment by me2beats** [comment]
This would also make it much easier to write plugins that find nodes in the editor tree.
It would definitely be much more reliable than `get_child(0).get_child(1)..` or even `get_child_by_types(["VBoxContainer", "HBoxContainer"])`.

Sometimes finding a node in a tree is the only way to get a node.

**Comment by Calinou** [comment]
> This would also make it much easier to write plugins that find nodes in the editor tree.

As I said in the proposal OP, names would not be guaranteed to remain consistent across versions since these are just intended to be debugging helpers. In practice, using names will likely break just as often as using `get_child()` 🙂

**Comment by Calinou** [comment]
> Is this proposal something agreed upon/vetted by the dev team? 

It's a bit controversial; see discussion in https://github.com/godotengine/godot/pull/76563.

**Comment by NathanLovato** [comment]
Some editor nodes already have a name or a type that can be used to locate them even if they are not exposed as part of the plugin API. Have you had issues with people accessing them and telling you they shouldn't change?

I would expect people to find it reasonable that anything that's not exposed in the public plugin API is not supported.

As an end user, in your Godot projects, you tend to move and rename nodes a lot throughout a project's lifetime, so it should be understandable that the editor is the same and that node changes are needed to improve it.

**Comment by YuriSizov** [comment]
Having had worked with Nathan on a library that needs to traverse the editor tree with a certain degree of confidence, I'm now of an opinion that naming of internal nodes is an important tool for debugging the editor itself. I do not mean to retract my own take that doing this would create an impression of internal stability. However, to me now this is similar to the perceived stability of named nodes in our own project scenes. We give names to reflect purpose, and sometimes type, of the node, and as it changes the name can change too. Same if the composition changes as well.

To that end, the focus should be less on giving things unique names and more on giving them contextually relevant names for their purpose. Some parts of the editor already do that, though most don't. Around the parts that do, the benefits when debugging the layout are immediately obvious though. You don't need to look at the source code or highlight nodes via some plugin to understand which part of the layout it is supposed to be. It's even more glaring when you have one named node and a few unnamed next to it.

Giving key parts of various widgets and editors meaningful names can also expose issues with logical separation. You can look no further than 2D and 3D editors. They follow the same structure, and while 3D editor is pretty well organized, naming and separation of meaningful parts in the 2D editor is all over the place. Things being called a viewport when they are not, that kind of stuff. While not immediately useful to most users, or even most plugin developers, one can argue, it is a matter of the internal code quality. I can guess that certain names predate current layout/composition structure, but then it's time to catch up and name types to remove confusion or misdirection.

Finally, there are also quirks. Like, there are nodes in the TileMap/TileSet editors which get names from translated strings, so node names themselves are translated, which is pretty bad. Node names need to be consistent regardless of the language, because they form a formal structure. (There is also a matter of TTR and TTRC being used inconsistently in that particular case, leading to some TileMap nodes being translated by the editor utilities, and some by auto-translation of Controls.)

-----

So, TL;DR is this. I believe that each editor widget/component should be reviewed as if you're creating a scene in Godot, and meaningful parts should be given meaningful, but local names. You should be able to inspect the tree and get a quick understanding that this is a toolbar, this is the content area, and this is an error label. Stuff like that. No need to name every single node. And no need to worry about layout changes affecting plugin devs. But do need to worry about updating names according to layout changes, of course.

**Comment by ryevdokimov** [comment]
I agree with the last things that have been said and have had similar experiences. I've started work on it above.
